File: cpp/713.subarray-product-less-than-k.cpp
/*
 * @lc app=leetcode id=713 lang=cpp
 *
 * [713] Subarray Product Less Than K
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numSubarrayProductLessThanK(vector<int> &nums, int k)
    {
        if (k <= 1)
        {
            return 0;
        }
        int res = 0;
        int width = nums.size();
        int left = 0;
        int cum_val = 1;
        for (int right = 0; right != width; ++right)
        {
            cum_val *= nums[right];
            while (cum_val >= k)
            {
                cum_val /= nums[left++];
            }
            res += right - left + 1;
        }
        return res;
    }
};
// @lc code=end

File: py/713.subarray-product-less-than-k.py
#
# @lc app=leetcode id=713 lang=python3
#
# [713] Subarray Product Less Than K
#

# @lc code=start
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        res = 0
        width = len(nums)
        left = 0
        cum_val = 1
        for right in range(width):
            cum_val *= nums[right]
            while cum_val >= k:
                cum_val //= nums[left]
                left += 1
            res += right-left+1
        return res
# @lc code=end
